lib/ansible/modules/cloud/amazon/ec2_vpc_igw.py

```diff
@@ -73,57 +80,103 @@ class AnsibleIGWException(Exception):
 
-    for igw in igws:
-        try:
-            vpc_conn.detach_internet_gateway(igw.id, vpc_id)
-            vpc_conn.delete_internet_gateway(igw.id)
-        except EC2ResponseError as e:
-            raise AnsibleIGWException(
-                'Unable to delete Internet Gateway, error: {0}'.format(e))
 
-    return {'changed': True}
+def ensure_tags(vpc_conn, resource_id, tags, add_only, check_mode):
+    try:
+        cur_tags = get_resource_tags(vpc_conn, resource_id)
+        if cur_tags == tags:
+            return {'changed': False, 'tags': cur_tags}
+
+        to_delete = dict((k, cur_tags[k]) for k in cur_tags if k not in tags)
+        if to_delete and not add_only:
+            vpc_conn.delete_tags(resource_id, to_delete, dry_run=check_mode)
 
+        to_add = dict((k, tags[k]) for k in tags if k not in cur_tags or cur_tags[k] != tags[k])
+        if to_add:
+            vpc_conn.create_tags(resource_id, to_add, dry_run=check_mode)
 
-def ensure_igw_present(vpc_conn, vpc_id, check_mode):
-    igws = vpc_conn.get_all_internet_gateways(
-        filters={'attachment.vpc-id': vpc_id})
+        latest_tags = get_resource_tags(vpc_conn, resource_id)
+        return {'changed': True, 'tags': latest_tags}
+    except EC2ResponseError as e:
+        raise AnsibleTagCreationException(
+            'Unable to update tags for {0}, error: {1}'.format(resource_id, e))
 
+
+def get_matching_igw(vpc_conn, vpc_id):
+    igws = vpc_conn.get_all_internet_gateways(filters={'attachment.vpc-id': vpc_id})
     if len(igws) > 1:
         raise AnsibleIGWException(
             'EC2 returned more than one Internet Gateway for VPC {0}, aborting'
             .format(vpc_id))
+    return igws[0] if igws else None
 
-    if igws:
-        return {'changed': False, 'gateway_id': igws[0].id}
-    else:
+
+def ensure_igw_absent(vpc_conn, vpc_id, tags, check_mode):
+    igw = get_matching_igw(vpc_conn, vpc_id)
+    if igw is None:
+        return {'changed': False}
+
+    if check_mode:
+        return {'changed': True}
+
+    try:
+        vpc_conn.detach_internet_gateway(igw.id, vpc_id)
+        vpc_conn.delete_internet_gateway(igw.id)
+    except EC2ResponseError as e:
+        raise AnsibleIGWException(
+            'Unable to delete Internet Gateway, error: {0}'.format(e))
+
+    return {'changed': True}
+
+
+def ensure_igw_present(vpc_conn, vpc_id, tags, check_mode):
+    igw = get_matching_igw(vpc_conn, vpc_id)
+    changed = False
+    if igw is None:
         if check_mode:
             return {'changed': True, 'gateway_id': None}
 
         try:
             igw = vpc_conn.create_internet_gateway()
             vpc_conn.attach_internet_gateway(igw.id, vpc_id)
-            return {'changed': True, 'gateway_id': igw.id}
+            changed = True
         except EC2ResponseError as e:
             raise AnsibleIGWException(
                 'Unable to create Internet Gateway, error: {0}'.format(e))
 
+    igw.vpc_id = vpc_id
+
+    if tags != igw.tags:
```

Comment:
@s-hertel tags defaults to an empty dict as per the argument spec; can you mark this as reviewed?